addons/l10n_se/models/res_partner_bank.py

```diff
@@ -0,0 +1,43 @@
+# -*- coding: utf-8 -*-
+import re
+
+from odoo import api, fields, models,  _
+from odoo.exceptions import ValidationError
+
+# TODO: Move Luhn-algorithm to utility class
+def validate_luhn(acc_number):
+    n = len(acc_number)
+    digits = [int(d) for d in re.sub(r'\D', '', acc_number)][-n:]
+    even_digitsum = sum(x if x < 5 else x - 9 for x in digits[::2])
+    return 0 == sum(digits, even_digitsum) % 10
+
+
+class ResPartnerBank(models.Model):
+    _inherit = 'res.partner.bank'
+
+    @api.model
+    def _get_supported_account_types(self):
+        rslt = super(ResPartnerBank, self)._get_supported_account_types()
+        rslt.append(('plusgiro', _('Plusgiro')))
+        rslt.append(('bankgiro', _('Bankgiro')))
+
+        return rslt
+
+    @api.model
+    def retrieve_acc_type(self, acc_number):
+        if acc_number and re.match('\d{5,7}-\d{1}', acc_number):
+            return 'plusgiro'
+        elif acc_number and re.match('\d{3,4}-\d{4}', acc_number):
+            return 'bankgiro'
+        else:
+            super(ResPartnerBank, bank).retrieve_acc_type(acc_number)
+
+    @api.constrains('acc_number')
+    def acc_number_constrains(self):
+        for record in self:
+            if record.acc_type == 'plusgiro':
+                if not validate_luhn(record.acc_number):
+                    raise ValidationError(_('The Plusgiro account is not correct.'))
+            elif record.acc_type == 'bankgiro':
```

Comment:
Both branches of the `if elif` are the same, they can be merged, or the statement can be removed altogether if there are no other value for `record.acc_type`.

Reply:
Well they can be rewritten so they use the same statement. The only difference is actually the error message. So the statement isn't exactly the same. If you want to make the code shorter then you could rewrite the error message and just do one if check with an or statement. However what you are doing at this moment is suboptimizing. It doesn't improve code, speed or do anything useful. What the code brings at the moment is transperancy for the end user. If you rewrite it it will get lost.

Reply:
I understand, it's just that my opinion differs. I think that a statement such as
```
if not validate_luhn(record.acc_number):
    raise ValidationError(_(f'The {record.acc_type} account is not correct.'))
```
is **easier to read**. My first thought when I read the code was "Why are there two branches ? How is the account type relevant to the Luhn check?". Hence my remark.

Anyway I think it's down to your own preference. So I won't argue, I'm nit picking.

Reply:
```
    @api.constrains('acc_number')
    def acc_number_constrains(self):
        for record in self:
            if record.acc_type == 'plusgiro' or record.acc_type == 'bankgiro':
                if not check_valid_acc_number(record.acc_number):
                    raise ValidationError(_('The account number for %s is not correct.') % record.acc_type)
```